elastic/kibana
# [ResponseOps] implement task claiming strategy mget

resolves: https://github.com/elastic/kibana/issues/181325

## Summary

Adds a new task claiming strategy `mget`, which can be used instead of the default one `default`.  Add the following to your `kibana.yml` to enable it:

    xpack.task_manager.claim_strategy: 'unsafe_mget'

## Work still to be done

- #184937 - fix limited concurrency starvation of unlimited tasks (separate query as mquery?)
- #184938 - fix unrecognized task starvation of unlimited tasks (separate query as mquery?)
- #184940 - in some cases, this code returns no tasks as they are all stale/conflicted, should retry ...
- #184941 - ensure we can switch between claimers with a reboot
- #184942 - complete jest and function tests - the new claimer should have a full jest set of tests like the default claimer, it only has a few right now; a new function test config has been added, with the same tests from the default claimer, but has 4 skipped tests (2 constant, 2 flaky)
- #184943 - collect stale / conflict counts for metrics/etc, where to publish?
- #184946 - provide way to switch between claimers live

## To Verify

Add  `xpack.task_manager.claim_strategy: 'unsafe_mget'` to your `kibana.dev.yml` and you'll be using the new claimer.  The most obvious problem is the task manager health report has null/0 values for the task claim section (not sure why, yet). 

You can set TM debug logging on, and should see a message every claim cycle:

```
task claimer claimed: N; stale: N; conflicts: N; missing: N; updateErrors: N; removed: N;
```

Here's the logging stanza:

```
logging:
  loggers:
    - name:  plugins.taskManager
      level: debug
```

When running multiple Kibana instances, you'll see cases where all the potential tasks are stale, or the updates were in conflict, and nothing actually gets claimed (noted in things to fix, above).

If you want to dig a little further,  a command-line tool is available in `x-pack/plugins/task_manager/server/manual_tests/get_rule_run_event_logs.js`.  It is used to pull rule run execution documents from multiple clusters at once, and provide some augmented info in them, regarding workers, idle time, etc.  The idea is to do an A/B test of using the new task claimer vs default, then see how the runs compare.
